### PR TITLE
TST: corrwith and tshift in groupby/groupby.transform

### LINKED ISSUE #27905: Functions allowed in DataFrameGroupby.transform are not all tested
e.g. `fillna`, `tshift`and `corrwith` are all allowed in `group.transform()`, but looking at `pandas/tests/groupby/test_transform.py`, they are not tested. There are more, and I will add to here and make a PR to tests these cases.

### PR DESCRIPTION
**Pull request by ryankarlos**
This PR tests `corrwith` and `tshift`, so not sure if it fully closes the original issue or not, but the other functions`backfill`, `pad` and `cumcount` don't seem to be supported in `groupby.transform` as reported in issues  #27472 and #31269.

- [x] closes #27905 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
@ryankarlos is there more to add. if not I don't think this closes the issue?

**Comment by ryankarlos** [comment]
@simonjayhawkins  yes Im thinking of  adding for `tshift` and `corrwith` -  but theres loads more like `cumsum` , `cumcount` , `ngroup` etc - do these all need to be tested in this PR or can they be split as they are not explicitly mentioned in the original issue ?

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins yes Im thinking of adding for `tshift` and `corrwith` 

those are explicitly mentioned in the issue, so could add here or seperate PRs and change the OP of this PR to xref instead of closes.

> - anymore ?

not sure. @charlesdong1991 ?

**Comment by WillAyd** [comment]
So there is already a fixture called `transformation_func` which has all of these; I think should be using that in tests where there is a gap today instead of adding this

**Comment by ryankarlos** [comment]
Thanlks @WillAyd , Ive now added new tests with the `transformation_func` fixture.

File: pandas/tests/groupby/test_transform.py [hunk 1]
```diff
@@ -1196,3 +1196,61 @@ def test_transform_lambda_indexing():
...
+
+    df1 = DataFrame(
+        {
+            "A": [121, 121, 121, 121, 231, 231, 676],
+            "B": [1.0, 2.0, 2.0, 3.0, 3.0, 3.0, 4.0],
+        }
+    )
+    g1 = df1.groupby("A")
+
+    if transformation_func == "corrwith":
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: this is only testing these 3 functions, so just split this into 3 tests, much easier to understand. you can make the input df a fixture in this test module

File: pandas/tests/groupby/test_transform.py [hunk 2]
```diff
@@ -1196,3 +1196,61 @@ def test_transform_lambda_indexing():
...
+        result = df2.set_index("C").groupby("A").tshift(2, "D")
+        df2["C"] = dt_periods + dt_periods.freq * 2
+        expected = df2
+        tm.assert_frame_equal(
+            result.reset_index().reindex(columns=["A", "B", "C"]), expected
+        )
+
+
+def test_check_original_and_transformed_index(transformation_func):
+    df = DataFrame({"A": [0, 0, 0, 1, 1, 1], "B": [0, 1, 2, 3, 4, 5]})
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: add the issue number as a comment

File: pandas/tests/groupby/test_transform.py [hunk 3]
```diff
@@ -1196,3 +1196,61 @@ def test_transform_lambda_indexing():
         ),
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_transform_nan_tshift_corrwith(transformation_func):
+
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: add the issue number as a comment

File: pandas/tests/groupby/test_transform.py [hunk 4]
```diff
@@ -1196,3 +1196,61 @@ def test_transform_lambda_indexing():
         ),
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_transform_nan_tshift_corrwith(transformation_func):
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: this test name is not that informative

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by ryankarlos** [comment]
@jreback  thanks for the comments - addressed them now.

File: pandas/tests/groupby/test_transform.py [hunk 5]
```diff
@@ -1196,3 +1206,81 @@ def test_transform_lambda_indexing():
         ),
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_corrwith(transformation_func, df_for_transformation_func):
+
+    # GH 27905
+    df = df_for_transformation_func.copy()
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_transform.py]: is a copy needed? By default, the dataframe fixture is reevaluated for each test unless scope is passed to fixture decorator. see https://docs.pytest.org/en/latest/fixture.html#scope-sharing-a-fixture-instance-across-tests-in-a-class-module-or-session

File: pandas/tests/groupby/test_transform.py [hunk 6]
```diff
@@ -1196,3 +1206,81 @@ def test_transform_lambda_indexing():
         ),
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_corrwith(transformation_func, df_for_transformation_func):
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_transform.py]: if the test is for corrwith only, don't need to use transformation_func fixture. same for other tests.

**Review by simonjayhawkins — COMMENTED** [review]
@ryankarlos there is a test, test_transform_transformation_func on L321 that was added recently. Should these be included there or at least co-located with that test?

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins yes Im thinking of adding for `tshift` and `corrwith` - anymore ?

looking at 

https://github.com/pandas-dev/pandas/blob/53ece7009c02eefc0befccf911d72369b0ed37f8/pandas/tests/groupby/test_transform.py#L330-L332

it appears that pad and cumcount need to be tested also.

**Comment by ryankarlos** [comment]
> > @simonjayhawkins yes Im thinking of adding for `tshift` and `corrwith` - anymore ?
> 
> looking at
> 
> https://github.com/pandas-dev/pandas/blob/53ece7009c02eefc0befccf911d72369b0ed37f8/pandas/tests/groupby/test_transform.py#L330-L332
> 
> it appears that pad and cumcount need to be tested also.

So these were throwing `AttributeErrors` when used with transform which are also mentioned in #31269 and #27472

File: pandas/tests/groupby/test_transform.py (hunk 6 above)
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: I didnt see the existing `test_transform_transformation_func` - i guess I could merge all these tests in that one function  - although some of these are just `groupby` tests so not sure if its better to have them separate - could push them up though ?

File: pandas/tests/groupby/test_transform.py (hunk 5 above)
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: thanks, yeh your right -  i had assumed default scope  was `session` - will update.

File: pandas/tests/groupby/test_transform.py [hunk 7]
```diff
@@ -346,6 +356,79 @@ def test_transform_transformation_func(transformation_func):
...
+def test_groupby_tshift(df_for_transformation_func):
+
+    # GH 27905
+    df = df_for_transformation_func
+    dt_periods = pd.date_range("2013-11-03", periods=7, freq="D")
+    df["C"] = dt_periods
+    g = df.set_index("C").groupby("A")
+
+    op = lambda x: getattr(x, "tshift")(2, "D")
+    result = op(g)
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_transform.py]: g.tshift(...) for clarity? if not tested elsewhere

File: pandas/tests/groupby/test_transform.py [hunk 8]
```diff
@@ -346,6 +356,79 @@ def test_transform_transformation_func(transformation_func):
...
+    df["B"] = [1, np.nan, np.nan, 3, np.nan, 3, 4]
+    result = g.transform("fillna", value=1)
+    expected = pd.DataFrame({"B": [1.0, 1.0, 1.0, 3.0, 1.0, 3.0, 4.0]})
+    tm.assert_frame_equal(result, expected)
+    op = lambda x: getattr(x, "fillna")(1)
+    result = op(g)
+    tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_tshift(df_for_transformation_func):
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_transform.py]: should this test be testing `g.transform("tshift", ...)`?

File: pandas/tests/groupby/test_transform.py [hunk 9]
```diff
@@ -346,6 +356,79 @@ def test_transform_transformation_func(transformation_func):
...
+    g = df.groupby("A")
+
+    op = lambda x: getattr(x, "corrwith")(df)
+    result = op(g)
+    expected = pd.DataFrame(dict(B=[1, np.nan, np.nan], A=[np.nan] * 3))
+    expected.index = pd.Index([121, 231, 676], name="A")
+    tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_transform_nan(df_for_transformation_func):
```
[the remark this answers is not in the document]
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: Yes sorry i missed that - ive removed this test now as its redundant

File: pandas/tests/groupby/test_transform.py (hunk 8 above)
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: The output of this seems to be buggy - ive just raised an issue #32344

File: pandas/tests/groupby/test_transform.py [hunk 10]
```diff
@@ -346,6 +356,79 @@ def test_transform_transformation_func(transformation_func):
...
+    op = lambda x: getattr(x, "tshift")(2, "D")
+    result = op(g)
+    df["C"] = dt_periods + dt_periods.freq * 2
+    expected = df
+    tm.assert_frame_equal(
+        result.reset_index().reindex(columns=["A", "B", "C"]), expected
+    )
+
+
+def test_check_original_and_transformed_index(transformation_func):
```
[the remark this answers is not in the document]
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: Ive removed this now

File: pandas/tests/groupby/test_transform.py [hunk 11]
```diff
@@ -346,6 +356,79 @@ def test_transform_transformation_func(transformation_func):
...
 
 
+def test_groupby_corrwith(df_for_transformation_func):
+
+    # GH 27905
+    df = df_for_transformation_func
+    g = df.groupby("A")
+
+    op = lambda x: getattr(x, "corrwith")(df)
+    result = op(g)
```
[the remark this answers is not in the document]
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: so the only instance using `corrwith` in `test_groupby.py` is in `test_dup_labels_output_shape` which doesn't seem to be fully testing the output of `corrwith` as is done here

File: pandas/tests/groupby/test_transform.py [hunk 12]
```diff
@@ -346,6 +358,47 @@ def test_transform_transformation_func(transformation_func):
...
+
+    result = g.corrwith(df)
+    expected = pd.DataFrame(dict(B=[1, np.nan, np.nan], A=[np.nan] * 3))
+    expected.index = pd.Index([121, 231, 676], name="A")
+    tm.assert_frame_equal(result, expected)
+
+    with pytest.raises(AttributeError) as m:
+        g.transform("corrwith", df)
+
+    m.match("'Series' object has no attribute 'corrwith'")
```
Comment (MomIsBestFriend) [inline pandas/tests/groupby/test_transform.py]: The pattern the we usually do is:

```python
msg = "Foo bar baz"
with pytest.raises(AttributeError, match=msg):
    # some code
```

**Review by ShaharNaveh — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_transform.py (hunk 12 above)
Comment (ryankarlos) [reply pandas/tests/groupby/test_transform.py]: ok changed this

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
A quick glance, it appears that with the latest changes, the tests are now not testing for expected behavior, since there are bugs that need to be fixed first. Is that correct @charlesdong1991 ?

**Comment by ryankarlos** [comment]
@simonjayhawkins  to be honest, think this original issue may be out of date and superseded by #32293 ,  #31269 , #27472 and #32344 reporting the associated bugs which need to be fixed whilst adding the tests.

**Comment by simonjayhawkins** [comment]
> otherwise feel free to close this PR if not relevant anymore ?

Thanks @ryankarlos for the work here, much appreciated, but I'll close this PR and close the issue too.

**Pull request closed without merging**